src/v/transform/rpc/service.cc

```diff
@@ -174,13 +176,27 @@ ss::future<result<model::offset, cluster::errc>> local_service::produce(
 
 ss::future<result<stored_wasm_binary_metadata, cluster::errc>>
 local_service::store_wasm_binary(
-  iobuf data, model::timeout_clock::duration timeout) {
+  model::wasm_binary_iobuf data, model::timeout_clock::duration timeout) {
     uuid_t key = uuid_t::create();
     storage::record_batch_builder b(
       model::record_batch_type::raft_data, model::offset(0));
     std::vector<model::record_header> headers;
     headers.push_back(make_header("state", "live"));
-    b.add_raw_kw(make_iobuf(key), std::move(data), std::move(headers));
+    if (data().get_owner_shard() == ss::this_shard_id()) {
+        // If we're on the same shard we can enable move fragment optimizations
+        b.add_raw_kw(make_iobuf(key), std::move(*data()), std::move(headers));
+    } else {
+        // Otherwise we can't touch the memory and we need to do a full copy to
+        // this core.
```

Comment:
Yeah otherwise another core will touch the reference counters, which is maybe safe, but we've ran into bugs here with: https://github.com/redpanda-data/redpanda/issues/13447#issuecomment-1752007143 and https://github.com/redpanda-data/redpanda/pull/14058